runtime/debugger.c

```diff
@@ -440,7 +481,13 @@ void caml_debugger(enum event_kind event, value param)
       break;
     case REQ_CHECKPOINT:
 #ifndef _WIN32
+      caml_release_domain_lock (); /* Don't fork while holding locks. */
+      Unlock (dbg_out);
+      Unlock (dbg_in);
       i = fork();
+      Lock (dbg_in);
+      Lock (dbg_out);
+      caml_acquire_domain_lock ();
```

Comment:
Comparing this code with `caml_atfork_default` in domain.c, maybe you need to call `caml_reset_domain_lock` before re-acquiring the lock?